### PR TITLE
Fix #858 - Remove request library

### LINKED ISSUE #858: Replacement of the deprecated request library
Hi,

Starting to work on the Sigfox IoT Agent, I noticed it uses the [request](https://github.com/request/request) library, which is marked as fully deprecated as of Feb 2020 (https://github.com/request/request#deprecated). And so does the IoT Agent node lib.

Has this been spotted? Has a replacement already been identified?

FWIW, there is a list of alternatives here: https://github.com/request/request/issues/3143

### PR DESCRIPTION
**Pull request by jason-fox**
- Replace `request` with `got`
- Add shim to mimic the old request callback interface
- Expose shim in API (for use downstream)
- Update documentation
- Update tests (replace `request` with a utility function)

### CODE REVIEW

**Review by fgalan — APPROVED** [review]
LGTM

Waiting for LGTM on dependand PRs before to run a CI e2e integration test and merge. Pending:

- [x] LGTM on https://github.com/telefonicaid/iotagent-json/pull/594
- [x] LGTM on https://github.com/telefonicaid/iotagent-ul/pull/515
- [ ] LGTM on https://github.com/telefonicaid/lightweightm2m-iotagent/pull/255
- [x] LGTM on https://github.com/telefonicaid/sigfox-iotagent/pull/113
- [ ] LGTM on https://github.com/telefonicaid/iotagent-manager/pull/258

**Comment by jason-fox** [comment]
It looks like the four dependent PRs have been approved. Does anything else need to be done before you can merge?

-  telefonicaid/iotagent-json#594 
-  telefonicaid/iotagent-ul#515 
-  telefonicaid/sigfox-iotagent#113
-  telefonicaid/lightweightm2m-iotagent#255

**Comment by fgalan** [comment]
> It looks like the four dependent PRs have been approved. Does anything else need to be done before you can merge?
> 
>     * [Switch from request to request-shim iotagent-json#594](https://github.com/telefonicaid/iotagent-json/pull/594)
> 
>     * [Switch from request to request-shim iotagent-ul#515](https://github.com/telefonicaid/iotagent-ul/pull/515)
> 
>     * [Update to use request shim. sigfox-iotagent#113](https://github.com/telefonicaid/sigfox-iotagent/pull/113)
> 
>     * [Update to use request shim. lightweightm2m-iotagent#255](https://github.com/telefonicaid/lightweightm2m-iotagent/pull/255)

Yes. A cuestion from @mapedraza is pending here: https://github.com/telefonicaid/iotagent-manager/pull/258#discussion_r738991595

**Comment by jason-fox** [comment]
Response here: https://github.com/telefonicaid/iotagent-manager/pull/258/files/04ad57b7260a6e17e0ff43bfa8feb14a8501658b#r773869067

File: CHANGES_NEXT_RELEASE
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:2]: Maybe these lines correspond to a wrong solved merge conflict and should be recovered?

Comment (jason-fox) [reply CHANGES_NEXT_RELEASE]: Fixed 2b008d2

**Comment by fgalan** [comment]
After having a new look to all the involved PRs, all them are ready and test are passing at this point:

- #1119 at commit https://github.com/telefonicaid/iotagent-node-lib/pull/1119/commits/2b008d2212949163ed663dccde318291061557f2
-  telefonicaid/iotagent-json#594 at commit https://github.com/telefonicaid/iotagent-json/pull/594/commits/dfc9d1ff7f392dc3b28a58f5bbff727841d50125
-  telefonicaid/iotagent-ul#515 at commit https://github.com/telefonicaid/iotagent-ul/pull/515/commits/206b7a1e6c304558e9fa5ee1ccffaf4fe47eb55a
-  telefonicaid/sigfox-iotagent#113 at commit https://github.com/telefonicaid/sigfox-iotagent/pull/113/commits/0d3ceee7d1fc52c4561a2de3e0f59353e6e0bd0b
-  telefonicaid/lightweightm2m-iotagent#255 at commit https://github.com/telefonicaid/lightweightm2m-iotagent/pull/255/commits/733f0acfe70c2399b2403578c8dd0d44125af04a
- telefonicaid/iotagent-manager#258 at commit https://github.com/telefonicaid/iotagent-manager/pull/258/commits/b10dd5184b04255d261db0aa146e86e25720bdf4

So we can go with the merging plan:

1. Upgrade IOTA-UL, IOTA-JSON, IOTA-Sigfox, IOTA-LWM2M and IOTA Manager PRs to revert the change in packages.json and point back to master branch in library. This will cause test to break in these PRs, but no problem (@jason-fox )
2. Merge this PR (@fgalan )
3. Re-run test in IOTA-UL, IOTA-JSON, IOTA-Sigfox, IOTA-LWM2M and IOTA Manager PRs to check thy go back to green (@fgalan )
4. Merge IOTA-UL, IOTA-JSON, IOTA-Sigfox, IOTA-LWM2M and IOTA Manager PRs (@fgalan)
5. Check that CI e2e works with the new docker images and report the result (@fgalan )

Probably due to Christmas holidays some of these steps will take a time.

**Comment by jason-fox** [comment]
@fgalan - I've broken all the unit tests for you.  Node 10 is now failing with a nasty [memory leak](https://stackoverflow.com/questions/56354079/npm-build-gives-ineffective-mark-compacts-near-heap-limit-allocation-failed) issue.

**Comment by fgalan** [comment]
> @fgalan - I've broken all the unit tests for you. Node 10 is now failing with a nasty memory leak issue.

I have merged this PR, then relaunch tests on IOTAS's PRs. All them are failing...

I'm seeing this kind of weird behaviour also in some unrelated repository such as https://github.com/telefonicaid/fiware-sth-comet/runs/4760353910?check_suite_focus=true or https://github.com/telefonicaid/perseo-fe/runs/4760670913?check_suite_focus=true

Why is this happening now? Maybe GitHub has reduce the resources allocated to GitHub Actions so the memory leaks now happens with a greater probability?

**Comment by fgalan** [comment]
> Node 12.x and 14.x are having the same issue as shown in the test results

Yes. This PR shows it: https://github.com/telefonicaid/iotagent-node-lib/pull/1171

Maybe it's an overall issue in Github Actions.

**Comment by fgalan** [comment]
All PRs has been merged. We are now in step 5. Let's see tomorrow how it goes (CI e2e runs nighly)

**Comment by fgalan** [comment]
Problems were detected in CI e2e but after the extra fix done in PR #1178 (and related PRs #1183 and https://github.com/telefonicaid/iotagent-json/pull/616) CI e2e tests are now passing so the work seems to be completed.

**Pull request merged**
